rule varscan_somatic:
    input:
        mpileup="mpileup/{sample}.mpileup.gz", # A pair of pileup files can be used *instead* of the mpileup
         # normal_pileup = ""
         # tumor_pileup = ""
    output:
        snp="vcf/{sample}.snp.vcf",
        indel="vcf/{sample}.indel.vcf",
    message:
        "Calling somatic variants {wildcards.sample}"
    threads: 1
    # optional specification of memory usage of the JVM that snakemake will respect with global
    # resource restrictions (https://snakemake.readthedocs.io/en/latest/snakefiles/rules.html#resources)
    # and which can be used to request RAM during cluster job submission as `{resources.mem_mb}`:
    # https://snakemake.readthedocs.io/en/latest/executing/cluster.html#job-properties
    resources:
        mem_mb=1024,
    params:
        extra="",
    wrapper:
        "master/bio/varscan/somatic"
